whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/ElasticFulltextSearch.java

```diff
@@ -112,50 +115,69 @@ protected SearchResponse doSearch() throws IOException {
                 final SearchHit[] hits = fulltextResponse.getHits().getHits();
 
                 LOGGER.debug("ElasticSearch {} hits for the query: {}", hits.length, searchRequest.getQuery());
-                final List<RpslObject> results = Lists.newArrayList();
-                int resultSize = Math.min(maxResultSize,Long.valueOf(fulltextResponse.getHits().getTotalHits().value).intValue());
+                return prepareResponse(fulltextResponse, hits, searchRequest, stopwatch);
+            }
+        }.search();
+    }
+
+    private SearchResponse prepareResponse(final org.elasticsearch.action.search.SearchResponse fulltextResponse,
+                                           final SearchHit[] hits,
+                                           final SearchRequest searchRequest,
+                                           final Stopwatch stopwatch) {
+        final List<SearchResponse.Lst> highlightDocs = Lists.newArrayList();
+        final List<SearchResponse.Result.Doc> resultDocumentList = Lists.newArrayList();
+        fillHighlightsAndResults(hits, highlightDocs, resultDocumentList);
 
-                final SearchResponse.Lst highlight = new SearchResponse.Lst("highlighting");
-                final List<SearchResponse.Lst> highlightDocs = Lists.newArrayList();
+        final SearchResponse.Result result = new SearchResponse.Result("response", Long.valueOf(fulltextResponse.getHits().getTotalHits().value).intValue(), searchRequest.getStart());
+        result.setDocs(resultDocumentList);
+
+        final SearchResponse searchResponse = new SearchResponse();
+        searchResponse.setResult(result);
 
-                for (final SearchHit hit : hits) {
-                    try {
-                        final RpslObject rpslObject = objectDao.getById(Integer.parseInt(hit.getId()));
-                        results.add(rpslObject);
-                        account(rpslObject);
+        final List<SearchResponse.Lst> responseLstList = Lists.newArrayList();
 
-                        highlightDocs.add(createHighlights(hit));
+        responseLstList.add(getResponseHeader(searchRequest, stopwatch.elapsed(TimeUnit.MILLISECONDS)));
 
```

Comment:
Why response header function needs a stopwatch as parameter?

Reply:
yes, it is weird. Thank you for pointing on this

Reply:
It seems that is not something new look master 
FulltextSearch.getResponseHeader
it is already like that in master...let me think how to refactor this...